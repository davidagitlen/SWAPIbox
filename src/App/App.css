@font-face {font-family: "STARWARS";
  src: url("//db.onlinewebfonts.com/t/c00548f5d217ed324ae9096051927001.eot");
  src: url("//db.onlinewebfonts.com/t/c00548f5d217ed324ae9096051927001.eot?#iefix") format("embedded-opentype"),
  url("//db.onlinewebfonts.com/t/c00548f5d217ed324ae9096051927001.woff") format("woff"),
  url("//db.onlinewebfonts.com/t/c00548f5d217ed324ae9096051927001.ttf") format("truetype"),
  url("//db.onlinewebfonts.com/t/c00548f5d217ed324ae9096051927001.svg#STARWARS") format("svg");
}

* {
  box-sizing: border-box;
}
.App {
  font-family: 'STARWARS';
  text-align: center;
}
