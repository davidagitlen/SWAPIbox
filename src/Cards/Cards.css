.Cards {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-around;
  height: 100vh;
  width: 100%;
  background: no-repeat center url(https://wallpaperplay.com/walls/full/b/e/4/124178.jpg);
  overflow: scroll;
}