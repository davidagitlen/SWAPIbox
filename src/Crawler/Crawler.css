.Crawler {
  color: yellow;
  background: repeat url(https://thumbs.gfycat.com/AccurateNimbleBighornedsheep.webp);
  padding: 15px;
  overflow: hidden;
  display: flex;
  justify-content: center;
  height: 1200px;
  perspective: 400px;
  font-size: 500%;
  font-weight: bold;
  letter-spacing: 6px;
  line-height: 150%;
  text-align: justify;
}

.subtext {
  text-align: center;
}

.crawl-text {
  position: relative;
  top: -100px;
  transform-origin: 50% 100%;
  animation: crawl 45s linear;
  animation-iteration-count: infinite;
}

.fade {
  position: absolute;
  width: 100%;
  min-height: 30vh;
  top: 210px;
  background-image: linear-gradient(0deg, transparent, black 90%);
  z-index: 1;
}

@keyframes crawl {
  0% {
    top: 300px;
    transform: rotateX(20deg) translateZ(0);
  }

  100% {
    top: -6000px;
    transform: rotateX(25deg) translateZ(-2500px);
  }

}