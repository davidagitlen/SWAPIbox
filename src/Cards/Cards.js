import React from 'react';
import PropTypes from 'prop-types';
import Card from '../Card/Card';
import './Cards.css';

const Cards = ({data, favoriteCard, favoritesArray}) => {
  let cardsList = data.map(datum => {
    const { name } = datum;
    return <Card key={name + Date.now()} id={name + Date.now()} data={datum} favoriteCard={favoriteCard} favoritesArray={favoritesArray} />
    }
  )

  return (
    <div className='Cards'>
      {cardsList}
    </div>
  )

}


export default Cards;


Cards.propTypes = {
  data: PropTypes.array,
  favoriteCard: PropTypes.func
}