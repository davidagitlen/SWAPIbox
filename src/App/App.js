import React, { Component }from 'react';
import './App.css';
import { Route, Navlink } from 'react-router-dom';
import Header from '../Header/Header';
import Cards from '../Cards/Cards';
import Crawler from '../Crawler/Crawler';


class App extends Component {
  constructor() {
    super();
    this.state = {
      vehicles: [],
      people: [],
      planets: [],
      crawl: {},
      favorites: [],
      isLoading: true
    }
  }

  componentDidMount() {
    fetch('https://swapi.co/api/people/')
      .then(response => response.json())
      .then(data => this.fetchPeople(data.results))
      .then(data => this.fetchSpecies(data))
      .then(people => this.setState({ people: people}))
      .catch(error => console.log(error));
      setTimeout(() => {
        console.log('state', this.state)
      }, 3000);

    fetch('https://swapi.co/api/planets/')
      .then(response => response.json())
      .then(data => this.fetchResidents(data.results))
      .then(planets => this.setState({ planets: planets}))
      .catch(error => console.log(error));

    fetch('https://swapi.co/api/vehicles/')
      .then(response => response.json())
      .then(data => this.fetchVehicles(data.results))
      .then(vehicles => this.setState({ vehicles: vehicles}))
      .catch(error => console.log(error))

    this.fetchFilm();

    if (localStorage.getItem('favorites')) {
      const favorites = JSON.parse(localStorage.getItem('favorites'));
      this.setState({ favorites }, () => {console.log('in the localstorage conditional', this.state.favorites)});
    }
  }

  fetchFilm = () => {
    let filmNumber = Math.floor(Math.random() * (7 -1) + 1)
    fetch(`https://swapi.co/api/films/${filmNumber}`)
      .then(response => response.json())
      .then(film => ({title: film.title, crawl: film.opening_crawl, release: film.release_date}))
      .then(crawl => this.setState({ crawl: crawl, isLoading: false }))
      .catch(error => console.log(error))
  }

  fetchPeople = (people) => {
    const promises = people.map(person => {
      return fetch(person.homeworld)
        .then(response => response.json())
        .then(data => ({
          favorite: 'false',
          type: 'people',
          homeworld: data.name,
          population: data.population,
          name: person.name,
          species: person.species
         }))
        .catch(error => console.log(error));
    });
    return Promise.all(promises);
  };

  fetchSpecies = (people) => {
    const promises = people.map(person => {
      return fetch(person.species)
        .then(response => response.json())
        .then(data => ({
          ...person,
          language: data.language,
          species: data.name
        }))
        .catch(error => console.log(error));
    });
    return Promise.all(promises);
  };

  fetchResidents = (planets) => {
    const planetsArray = planets.map(planet => {
      let nameArray = [];
      planet.residents.map(resident => {
        return fetch(resident)
        .then(response => response.json())
        .then(data => nameArray.push(data.name))
        .catch(error => console.log(error))
      });
      return {
        favorite: 'false',
        type: 'planets',
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: nameArray
      }
    });
    console.log('planets', planetsArray[0].residents.length)
    return planetsArray
  };

  fetchVehicles = (vehicles) => {
    return vehicles.map(vehicle => {
      return {
        favorite: 'false',
        type: 'vehicles',
        name: vehicle.name,
        model: vehicle.model,
        vehicleClass: vehicle.vehicle_class,
        numberOfPassengers: vehicle.passengers
      }
    });
  };

  // favoriteCard = (name, type, favorite) => {
  //   const foundCard = this.state[type].find(card => card.name === name);
  //   foundCard.favorite = foundCard.favorite === 'false' ? 'true' : 'false';
  //   if (foundCard.favorite === 'true' && !this.state.favorites.includes(foundCard)) {
  //     this.setState({ favorites: [...this.state.favorites, foundCard] }, () => { console.log('localstorage experiment toggle on', this.state.favorites)});
  //     this.setLocalStorage([...this.state.favorites, foundCard]);
  //   } else {
  //     const filteredFavorites = this.state.favorites.filter(card => card.favorite === 'true');
  //     this.setLocalStorage(filteredFavorites);
  //     this.setState({ favorites: filteredFavorites }, () => { console.log('localstorage experiment toggle off', this.state.favorites)});
      
  //   }
  // }



  favoriteCard = (cardData) => {
    console.log('in favoriteCard', cardData)
    const favoriteNames = this.state.favorites.map(favorite => favorite.name)

    if (!favoriteNames.includes(cardData.name)) {
      this.setLocalStorage([...this.state.favorites, cardData]);
      this.setState({favorites : [...this.state.favorites, cardData]}, () => {console.log(this.state.favorites)});
    } else {
      const filteredFavorites = this.state.favorites.filter(favorite => favorite.name !== cardData.name);
      this.setLocalStorage(filteredFavorites);
      this.setState({ favorites: filteredFavorites }, () => { console.log(this.state.favorites) });
    }
  }



  setLocalStorage = (favorites) => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }

  render() {
    return (
      <main className='App'>
        {this.state.isLoading && <p>Hold your horses</p>}
        <Header data={this.state.favorites}/>
        <Route exact path='/' render ={ () => <Crawler data={this.state.crawl} /> } />
        <Route exact path='/people' render={ () => <Cards data={this.state.people} favoriteCard={this.favoriteCard} favoritesArray={this.state.favorites} />} />
        <Route exact path='/planets' render={() => <Cards data={this.state.planets} favoriteCard={this.favoriteCard} favoritesArray={this.state.favorites}/>} />
        <Route exact path='/vehicles' render={() => <Cards data={this.state.vehicles} favoriteCard={this.favoriteCard} favoritesArray={this.state.favorites}/>} />
        <Route exact path='/favorites' render={() => <Cards data={this.state.favorites} favoriteCard={this.favoriteCard} favoritesArray={this.state.favorites}/>} />
      </main>
    )
  }
}

export default App;