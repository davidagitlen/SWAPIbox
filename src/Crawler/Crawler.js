import React from 'react';
import PropTypes from 'prop-types';
import './Crawler.css'

const Crawler = ({data}) => {
  const {crawl, title, release} = data;
  return(
    <>
    <div className='fade'></div>
    <section className='Crawler'>
      <div className='crawl-text'>
        <p>{crawl}</p>
        <p className='subtext'>{title}</p>
        <p className='subtext'>{release}</p>
      </div>
    </section>
    </>
  )
}

export default Crawler;

Crawler.propTypes = {
  data : PropTypes.object
}