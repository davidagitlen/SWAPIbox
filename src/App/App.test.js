import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import App from './App';
import Crawler from '../Crawler/Crawler';
import Cards from '../Cards/Cards';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />)
  });

  it('should render without crashing', () => {
    expect(wrapper.length).toEqual(1)
  });

  it('should update state to show favorited cards when favoriteCard is invoked', () => {
    const mockPerson = { name: 'David', homeworld: 'Kamino', species: 'wookie', population: 20, favorite: 'false' };

    const expected = [{ name: 'David', homeworld: 'Kamino', species: 'wookie', population: 20, favorite: 'true' }];

    wrapper.setState({ people: [mockPerson] });
    wrapper.instance().favoriteCard('David', 'people', 'true');

    expect(wrapper.state('people')).toEqual(expected);
    expect(wrapper.state('favorites')).toEqual(expected);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  describe('Route', () => {

    it('should route to the homepage with the Crawler displayed by default', () => {

      const wrapper = mount(
        <MemoryRouter initialEntries={['/']}>
          <App />
        </MemoryRouter>
      );

      expect(wrapper.find(Crawler)).toHaveLength(1);
    });

    it('should route to the people page when the people link is clicked', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/people']}>
          <App />
        </MemoryRouter>
      );
     
      expect(wrapper.find(Cards)).toHaveLength(1);
    });

    it('should route to the planets page when the planets link is clicked', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/planets']}>
          <App />
        </MemoryRouter>
      );

      expect(wrapper.find(Cards)).toHaveLength(1);
    });

    it('should route to the vehicles page when the vehicles link is clicked', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/vehicles']}>
          <App />
        </MemoryRouter>
      );

      expect(wrapper.find(Cards)).toHaveLength(1);
    });

    it('should route to the favorites page when the favorites link is clicked', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/favorites']}>
          <App />
        </MemoryRouter>
      );

      expect(wrapper.find(Cards)).toHaveLength(1);
    });
  });
});
