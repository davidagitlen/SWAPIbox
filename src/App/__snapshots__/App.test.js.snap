// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in correctly 1`] = `
<main
  className="App"
>
  <p>
    Hold your horses
  </p>
  <Header
    data={Array []}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    path="/people"
    render={[Function]}
  />
  <Route
    path="/planets"
    render={[Function]}
  />
  <Route
    path="/vehicles"
    render={[Function]}
  />
  <Route
    path="/favorites"
    render={[Function]}
  />
</main>
`;
